### ISSUE TITLE
Catch a page that runs AngularJS

### ISSUE THREAD

**Issue by MarioVillani** [ariya/phantomjs#13767]
I'm trying to scrape a page which is powered by AngularJS.
I'm using this script to get the contents:

```javascript
var page = new WebPage();
var fs = require('fs');
page.onLoadFinished = function(status) {
    fs.write('myfilename.html', page.content, 'w');
    phantom.exit();
}
page.open('www.myurl.com');
```
How can tell PhantomJS to wait for the rendering of all Angular elements in the page? I get the full rendered page only few times.

**Comment by rafis** [comment]
This is not issue of PhantomJS. Try asking on Stackoverflow.

Look at AngularJS application's code, look at controllers. After controller finishes initializing page it sets some attribute on `<body>` tag, for example, `<body data-status="ready">`. This is required for SEO.

**Comment by bnierop** [comment]
tl; dr; You need to wait for the page to render in memory before you rasterize to pdf/png


Example from rasterize.js, waiting for 2000ms
https://github.com/ariya/phantomjs/blob/master/examples/rasterize.js

page.open(address, function (status) {
        if (status !== 'success') {
            console.log('Unable to load the address!');
            phantom.exit(1);
        } else {
            window.setTimeout(function () {
                page.render(output);
                phantom.exit();
            }, 2000);
        }
    });

**Comment by MarioVillani** [comment]
I do not agree. The timeout leads to a race condition. Would be better to listen for a specific event sent from AngularJS that specifically says "I've finished rendering".

**Comment by TrejGun** [comment]
had same problem

on server
```
page.onConsoleMessage(message => {
	if (message.indexOf("!!!") !== -1) {
		// do what ever you need
	}
});
```

on page
```
Page.onAngularReady(function(){
	console.log("!!!");
})
```

**Comment by zackw** [comment]
This is a request for help, not a bug report.  Please only use this tracker to report problems that are *definitely* bugs in PhantomJS.

For general question, code review request, build problem, or other related discussion, please post it to either:

 * the official [PhantomJS users mailing-list](http://groups.google.com/group/phantomjs)
 * [Stack Overflow](http://stackoverflow.com/questions/tagged/phantomjs) (with `phantomjs` tag)

If you have not read it yet, please check also a few tips on [Effective Q&A](https://github.com/ariya/phantomjs/wiki/Support).
